README.md
# nTerminal
A multi-account ordering terminal written in C# to connect to the MT4 web version via WebSocket.
It be tested on Vultr's New York VPS, but is not guaranteed to be available now and in the future.
[Vultr VPS free $100 Credit to Get Started](https://www.vultr.com/?ref=8382011-6G)

#### 介绍
C#写的通过WebSocket连接MT4网页版的多账户下单终端。

把要登录的账号分master/slave一行一个账号(例：account,password,ICMarketsSC-Live06,master/slave)写在accounts.txt文件里，通过终端登录网页版MT4，终端即可操作slave账号跟着master账号下单/平仓。

终端与MT4官方服务器通过WebSocket直接通信。通信协议是根据网页版MT4的js文件和抓取WebSocket数据分析(猜)出来的。resources文件夹里是分析过程中的用的一些主要参考。

本人水平&精力有限很多东西并没有猜出来，but貌似不影响下单和平仓。

曾经实盘几十个账户在Vultr的New York虚拟机上跑了大概半年，但是时间久远不能保证现在及以后还能用。

[附：Vultr免费$100注册码](https://www.vultr.com/?ref=8382011-6G)

有兴趣的小伙伴可以一起研究研究 qq群：660937351


resources/history.txt
00 01 02 03 04 05 06 07  08 09 0A 0B 0C 0D 0E 0F 
--------------------------------------------------------------------
A0 7F 05 00 00 F0 DD 83  04 00 00 00 00 00 00 00   | ...............
00 00 00 00 00 00 00 00  00 06 00 00 00 01 00 00   | ................
00 C7 CB AB 5E 00 00 00  00 00 00 00 00 00 00 00   | ....^...........
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | ................
00 C7 CB AB 5E 00 00 00  00 00 00 00 00 00 00 00   | ....^...........
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | ................
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 40   | ...............@
6F C0 00 00 00 00 00 00  00 00 00 00 00 00 41 44   | o.............AD
43 55 50 2D 46 53 41 20  31 37 38 37 30 38 39 37   | CUP-FSA 17870897
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | ................
00 00 00 00 00 00 E8 D0  83 04 47 42 50 4A 50 59   | ..........GBPJPY
00 00 00 00 00 00 03 00  00 00 01 00 00 00 0A 00   | ................
00 00 26 B0 AB 5E 03 00  00 00 E1 7A 14 AE 47 D1   | ..&..^.....z..G.
60 40 00 00 00 00 00 00  00 00 33 33 33 33 33 C3   | `@........33333.
60 40 EC E6 AB 5E 00 00  00 00 00 66 66 66 66 66   | `@...^.....fffff
66 E6 BF 9A 99 99 99 99  99 C9 3F 00 00 00 00 00   | f.........?.....
00 00 00 91 ED 7C 3F 35  CE 60 40 71 3D 0A D7 A3   | .....|?5.`@q=...
F0 21 40 00 00 00 00 00  00 00 00 00 00 00 00 46   | .!@............F
58 77 68 74 7C 36 7C 37  35 37 34 37 35 35 36 7C   | Xwht|6|75747556|
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 F2   | ................
D2 4D 62 10 18 F4 3F FD  FE 83 04 47 42 50 4A 50   | .Mb...?....GBPJP
59 00 00 00 00 00 00 03  00 00 00 01 00 00 00 0A   | Y...............
00 00 00 34 F2 AB 5E 03  00 00 00 A2 45 B6 F3 FD   | ...4..^.....E...
CC 60 40 00 00 00 00 00  00 00 00 00 00 00 00 00   | .`@.............
00 00 00 34 FB AB 5E 00  00 00 00 00 66 66 66 66   | ...4..^.....ffff
66 66 E6 BF 9A 99 99 99  99 99 C9 3F 00 00 00 00   | ff.........?....
00 00 00 00 2D B2 9D EF  A7 CA 60 40 48 E1 7A 14   | ....-.....`@H.z.
AE 47 1B 40 00 00 00 00  00 00 00 00 00 00 00 00   | .G.@............
46 58 77 68 74 7C 36 7C  37 35 37 35 39 33 35 33   | FXwht|6|75759353
7C 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | |...............
32 3D 61 89 07 14 F4 3F  2C 28 84 04 47 42 50 4A   | 2=a....?,(..GBPJ
50 59 00 00 00 00 00 00  03 00 00 00 01 00 00 00   | PY..............
0A 00 00 00 08 09 AC 5E  03 00 00 00 7F 6A BC 74   | .......^....j.t
93 C8 60 40 00 00 00 00  00 00 00 00 33 33 33 33   | ..`@........3333
33 C3 60 40 4C 0B AC 5E  00 00 00 00 00 66 66 66   | 3.`@L..^.....fff
66 66 66 E6 BF 9A 99 99  99 99 99 C9 3F 00 00 00   | fff.........?...
00 00 00 00 00 6F 12 83  C0 CA C5 60 40 48 E1 7A   | .....o.....`@H.z
14 AE 47 20 40 00 00 00  00 00 00 00 00 00 00 00   | ..G @...........
00 46 58 77 68 74 7C 36  7C 37 35 37 36 39 38 39   | .FXwht|6|7576989
35 7C 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | 5|..............
00 62 F3 71 6D A8 18 F4  3F 77 2A 85 04 47 42 50   | .b.qm...?w*..GBP
4A 50 59 00 00 00 00 00  00 03 00 00 00 00 00 00   | JPY.............
00 0A 00 00 00 12 9C AF  5E 03 00 00 00 71 3D 0A   | ........^....q=.
D7 A3 A0 60 40 00 00 00  00 00 00 00 00 00 00 00   | ...`@...........
00 00 A8 60 40 13 A3 AF  5E 00 00 00 00 00 66 66   | ...`@...^.....ff
66 66 66 66 E6 BF 9A 99  99 99 99 99 C9 3F 00 00   | ffff.........?..
00 00 00 00 00 00 B6 F3  FD D4 78 9D 60 40 8F C2   | ..........x.`@..
F5 28 5C 8F 22 C0 00 00  00 00 00 00 00 00 00 00   | .(\."...........
00 00 46 58 77 68 74 7C  36 7C 37 35 38 33 36 30   | ..FXwht|6|758360
31 34 7C 00 00 00 00 00  00 00 00 00 00 00 00 00   | 14|.............
00 00 1E AC FF 73 98 EF  F3 3F FB 2E 85 04 47 42   | .....s...?....GB
50 4A 50 59 00 00 00 00  00 00 03 00 00 00 01 00   | PJPY............
00 00 0A 00 00 00 1A A3  AF 5E 03 00 00 00 F0 A7   | .........^......
C6 4B 37 9D 60 40 00 00  00 00 00 00 00 00 9A 99   | .K7.`@..........
99 99 99 99 60 40 94 A9  AF 5E 00 00 00 00 00 66   | ....`@...^.....f
66 66 66 66 66 E6 BF 9A  99 99 99 99 99 C9 3F 00   | fffff.........?.
00 00 00 00 00 00 00 C3  F5 28 5C 8F 9E 60 40 85   | .........(\..`@.
EB 51 B8 1E 85 0F C0 00  00 00 00 00 00 00 00 00   | .Q..............
00 00 00 46 58 77 68 74  7C 36 7C 37 35 38 33 37   | ...FXwht|6|75837
31 37 38 7C 00 00 00 00  00 00 00 00 00 00 00 00   | 178|............
00 00 00 20 24 0B 98 C0  ED F3 3F 59 42 85 04 47   | ... $.....?YB..G
42 50 4A 50 59 00 00 00  00 00 00 03 00 00 00 01   | BPJPY...........
00 00 00 0A 00 00 00 32  C4 AF 5E 03 00 00 00 BA   | .......2..^.....
49 0C 02 2B 9B 60 40 00  00 00 00 00 00 00 00 00   | I..+.`@.........
00 00 00 00 00 00 00 5D  C4 AF 5E 00 00 00 00 05   | .......]..^.....
66 66 66 66 66 66 E6 BF  9A 99 99 99 99 99 C9 3F   | ffffff.........?
00 00 00 00 00 00 00 00  FA 7E 6A BC 74 9B 60 40   | .........~j.t.`@
E1 7A 14 AE 47 E1 EA BF  00 00 00 00 00 00 00 00   | .z..G...........
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | ................
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00   | ................
00 00 00 00 EE 25 8D D1  3A EA F3 3F 78 42 85 04   | .....%..:..?xB..
47 42 50 4A 50 59 00 00  00 00 00 00 03 00 00 00   | GBPJPY..........
01 00 00 00 0A 00 00 00  71 C4 AF 5E 03 00 00 00   | ........q..^....
D7 A3 70 3D 0A 9B 60 40  33 33 33 33 33 A3 60 40   | ..p=..`@33333.`@
66 66 66 66 66 86 60 40  9A DA AF 5E 00 00 00 00   | fffff.`@...^....
00 66 66 66 66 66 66 E6  BF 9A 99 99 99 99 99 C9   | .ffffff.........
3F 00 00 00 00 00 00 00  00 B8 1E 85 EB 51 94 60   | ?............Q.`
40 EC 51 B8 1E 85 AB 33  40 00 00 00 00 00 00 00   | @.Q....3@.......
00 00 00 00 00 46 58 77  68 74 7C 36 7C 37 35 38   | .....FXwht|6|758
34 32 31 36 31 7C 00 00  00 00 00 00 00 00 00 00   | 42161|..........
00 00 00 00 00 7E 1D 38  67 44 E9 F3 3F 13 74 85   | .....~.8gD..?.t.
04 47 42 50 4A 50 59 00  00 00 00 00 00 03 00 00   | .GBPJPY.........
00 01 00 00 00 0A 00 00  00 3F E5 AF 5E 03 00 00   | .........?..^...
00 3F 35 5E BA 49 A4 60  40 00 00 00 00 00 00 00   | .?5^.I.`@.......

resources/wss.js
(function() {
    function y() {
        S.Wm && S.Wm.apply(S, arguments)
    }
    function x(b, a, c) {
        R && !window.td && (b = R[b]) && b.forEach(function(b) {
            b.call(S, a, c)
        })
    }
    function q(a) {
        var c = new DataView(a)
          , d = c.Yb(0, a.byteLength);
        a = ["__", String.fromCharCode(Math.ceil(25 * Math.random()) + 65)].join("");
        eval(["window.", a, " = ", d, ";"].join(""));
        d = String(window[a]);
        window[a] = null;
        delete window[a];
        a = new ArrayBuffer(d.length);
        c = new DataView(a);
        c.G(0, d);
        S.$e(b.Ok, a)
    }
    function g(b) {
        b = (new DataView(b)).getUint32(0, !0);
        u(C, b) && (b === C.sm && (y(!0, !1, "connection to trade server established", !0),
        (I === C.Cp || I === C.Bp) && S.Zm && S.Zm()),
        I = b,
        I === C.Cp && (y(!1, !1, "connection to trade server lost", !0, !0),
        n.se.ri()),
        I === C.Bp && (y(!1, !1, "authentication to trade server failed", !0, !0),
        n.se.ri()))
    }
    function u(b, a) {
        for (var c in b)
            if (b.hasOwnProperty(c) && b[c] === a)
                return !0;
        return !1
    }
    function w(a) {
        for (var c in b)
            if (b[c] === a)
                return !0;
        return !1
    }
    function p(b) {
        var a = b && b.byteLength || 0
          , c = new Uint8Array(8 + a);
        b && c.set(new Uint8Array(b), 8);
        c = c.buffer;
        b = new DataView(c);
        b.setUint32(0, a, !0);
        b.setUint32(4, 1, !0);
        M.yo++;
        M.send(b.buffer);
        Q.nk || (Q.nk = 0);
        Q.nk += b.buffer.byteLength
    }
    function k(a) {
        if (a) {
            var c = new DataView(a)
              , d = c.getUint16(2, !0)
              , e = c.getUint8(4, !0)
              , f = a.slice(5);
            M.G = d;
            window.Yb && d !== b.Lq && d !== b.hB && d !== b.eB && d !== b.gB && y(!0, !0, ["Receive command: ", d, " (size: ", a.byteLength, " bytes, header: ", c.getUint8(0, !0).toString(16), c.getUint8(1, !0).toString(16), ")"].join(""));
            if (d === b.Sk)
                f && f.byteLength && (S.LH = (new DataView(f)).getUint16(0, !0));
            else if (d === b.Ok)
                q(f);
            else if (d === b.bB)
                g(f);
            else {
                d === b.Qk && e === B.fb && n.$d && n.$d.Np(M.xh);
                if (0 < e) {
                    a = "";
                    switch (e) {
                    case 1:
                        a = "websocket returned internal error.";
                        break;
                    case 2:
                        a = "websocket returned error: trade server is not available.";
                        break;
                    case 3:
                        a = "websocket returned error: authorization error on the trading server."
                    }
                    y(!1, !0, a, !0)
                }
                x(d, e, f.byteLength ? f : null);
                w(d) || (M.yo = 0,
                setTimeout(function() {
                    M.close()
                }, 1E3))
            }
        }
    }
    function d(b) {
        if (b.data instanceof ArrayBuffer) {
            var a = b.data;
            (a = a.slice(8)) && n.Ud.ng(a, 1, k)
        }
        Q.hk || (Q.hk = 0);
        Q.hk += b.data.byteLength
    }
    function l() {
        y(!0, !0, "websocket connection is successfully established on", !0);
        M.removeEventListener("open", l);
        N && S.$e(b.Sk, N);
        if (window.La) {
            document.body.style.background = "green";
            try {
                window.parent && window.parent.postMessage && window.parent.postMessage(JSON.stringify({
                    status: "open"
                }), "http://localhost/")
            } catch (a) {}
        }
    }
    function r(a) {
        if (window.La) {
            document.body.style.background = "red";
            try {
                window.parent && window.parent.postMessage && window.parent.postMessage(JSON.stringify({
                    status: "close"
                }), "http://localhost/")
            } catch (c) {}
        } else
            n.se.ri(),
            M.G !== b.Mm && M.G !== b.Oq && M.G !== b.Qk && M.G !== b.fB && (y(a.cI ? !0 : !1, !0, a.cI ? "websocket connection was closed on the server side" : "websocket connection was suddenly closed", !0),
            y(!1, !1, "connection failed!", !0),
            window.td || (L ? (n.$d && n.$d.kd(),
            J ? J *= 2 : J = 1E3,
            J = Math.min(J, 6E4),
            S.Xm && setTimeout(function() {
                S.Xm(M)
            }, J)) : S.Ym && setTimeout(function() {
                S.Ym(M)
            }, 500)))
    }
    function e() {
        if (window.La) {
            document.body.style.background = "red";
            try {
                window.parent && window.parent.postMessage && window.parent.postMessage(JSON.stringify({
                    status: "error"
                }), "http://localhost/")
            } catch (b) {}
        } else
            y(!1, !0, "websocket connection reported an error!", !0),
            n.se.ri()
    }
    function z() {
        if (c.xb) {
            var b, a = window.localStorage.debug;
            a && a.indexOf && -1 !== a.indexOf("{") && (b = window.JSON.parse(a));
            b || (b = {});
            b.socket || (b.socket = {});
            b.socket["in"] = Q.hk;
            b.socket.out = Q.nk;
            window.localStorage.debug = window.JSON.stringify(b)
        }
    }
    function h() {
        S.$e(b.Oq)
    }
    var a = window.G, c = a.la, f = a.Z, t = a.wb, n = a.O, b = {
        Sk: 0,
        Qk: 1,
        Mm: 2,
        Jq: 3,
        Nm: 4,
        Lm: 5,
        Pm: 6,
        cB: 7,
        Lq: 8,
        Om: 9,
        dB: 10,
        Rk: 11,
        Mq: 12,
        Qm: 13,
        aB: 14,
        bB: 15,
        Pk: 16,
        GO: 21,
        Kq: 26,
        Pq: 0,
        Oq: 2,
        HO: 3,
        MO: 4,
        KO: 5,
        UO: 6,
        SO: 18,
        TO: 9,
        PO: 7,
        hB: 8,
        YO: 10,
        NO: 11,
        WO: 12,
        VO: 13,
        IO: 14,
        JO: 15,
        gB: 17,
        XO: 19,
        RO: 20,
        Rm: 21,
        QO: 22,
        eB: 23,
        LO: 24,
        OO: 25,
        Sm: 27,
        Tk: 30,
        fB: 28,
        Nq: 29,
        Ok: 50,
        PING: 51
    }, B = {
        fb: 0,
        GM: 1,
        sx: 2,
        rx: 3,
        HM: 4,
        EM: 5,
        FM: 6,
        tx: 7,
        ux: 8
    }, C = {
        sm: 0,
        Cp: 1,
        Bp: 2
    }, M, E, N, R, O, L = !0, J = 0, I = C.sm, P, Q = {}, S = n.Gb = {
        gq: 8,
        mb: b,
        ee: B,
        La: C,
        LH: null,
        Zm: null,
        Xm: null,
        Ym: null,
        Wm: null,
        aa: function(b) {
            E = b;
            window.Yb && (clearInterval(P),
            P = setInterval(z, 1E4));
            t.i(f.window, "beforeunload", h);
            return S
        },
        $e: function(a, c) {
            if (M && 1 === M.readyState && u(b, a)) {
                var d = c && c.byteLength || 0
                  , e = new Uint8Array(4 + d);
                c && e.set(new Uint8Array(c), 4);
                e = e.buffer;
                e = new DataView(e);
                e.setUint8(0, Math.floor(255 * Math.random()), !0);
                e.setUint8(1, Math.floor(255 * Math.random()), !0);
                e.setUint16(2, a, !0);
                n.Ud.Vd(e.buffer, 1, p, a === b.Sk || a === b.PING);
                window.Yb && y(!0, !0, ["Send command: ", a, " (", d, " bytes)"].join(""));
                return this
            }
        },
        WF: function(b) {
            b && (N = new ArrayBuffer(64),
            (new DataView(N)).G(0, b));
            return this
        },
        BC: function(b) {
            O = b;
            return this
        },
        Ur: function() {
            M && (2 > M.readyState && M.close(),
            M.removeEventListener("close", r),
            M.removeEventListener("message", d),
            M.removeEventListener("error", e));
            M = new WebSocket([!1 === O ? "ws://" : "wss://", E, "/"].join(""));
            M.binaryType = "arraybuffer";
            M.addEventListener("open", l);
            M.addEventListener("close", r);
            M.addEventListener("message", d);
            M.addEventListener("error", e);
            M.yo = 0;
            M.xh = E.split(".")[0];
            Q.hk = 0;
            Q.nk = 0;
            return M
        },
        gs: function() {
            L = !1;
            return this
        },
        Fd: function() {
            L = !0;
            J = 0;
            return this
        },
        zy: function() {
            return M && 1 === M.readyState
        },
        Ay: function() {
            return I === C.sm
        },
        Te: function(b, a, c) {
            R || (R = {});
            R[b] || (R[b] = []);
            a.Rc = (1E5 * Math.random()).toString();
            c = parseInt(c, 10);
            isNaN(c) ? R[b].push(a) : R[b].splice(c, 0, a)
        },
        Xk: function(b, a) {
            var c = a.Rc;
            if (!c || !R || !R[b])
                return !1;
            var d = R[b]
              , e = d.xt(function(b) {
                return b.Rc === c
            });
            return -1 < e ? (d.splice(e, 1),
            !0) : !1
        }
    }
}
)();